# 🧹 Introduction to Data Cleaning

## 🎯 Learning Objectives

- Understand what Data Cleaning is.
- Learn why Data Cleaning is essential.
- Identify common data quality issues.
- Understand the Data Cleaning workflow.
- Explore Data Cleaning using the Titanic dataset.

## 📖 Introduction

Data Cleaning is the process of identifying and correcting inaccurate, incomplete, duplicate, inconsistent, or irrelevant data. Clean data improves the quality of analysis, machine learning models, and business decisions.

## 🧹 What is Data Cleaning?

Data Cleaning is the process of preparing raw data by removing errors and inconsistencies to make it suitable for analysis and machine learning.

It helps improve:
- Accuracy
- Consistency
- Reliability
- Data Quality

## ❓ Why is Data Cleaning Important?

- Improves data quality.
- Increases model accuracy.
- Removes incorrect and duplicate records.
- Handles missing values.
- Makes data consistent and reliable.
- Saves time during analysis.

## ⚠️ Common Data Quality Issues

- Missing Values
- Duplicate Records
- Incorrect Data Types
- Outliers
- Inconsistent Formatting
- Invalid Values
- Incorrect Categorical Labels

## 🔄 Data Cleaning Workflow

1. Load the dataset.
2. Inspect the data.
3. Handle missing values.
4. Remove duplicates.
5. Correct data types.
6. Clean text data.
7. Detect and treat outliers.
8. Encode categorical variables.
9. Scale numerical features.
10. Save the cleaned dataset.

D:\Sanskruti\sanskruti document\Machine-Learning-From-Scratch\datasets\titanic.csv

In [6]:
import pandas as pd

df = pd.read_csv(r"D:\Sanskruti\sanskruti document\Machine-Learning-From-Scratch\datasets\train.csv")

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [7]:
df.tail()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q


In [8]:
df.shape

(891, 12)

In [9]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [10]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [12]:
df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [13]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [14]:
df.duplicated().sum()

np.int64(0)

## 🔍 Observations

From the Titanic dataset, we observed:

- The dataset contains **891 rows** and **12 columns**.
- Both numerical and categorical features are present.
- Columns such as **Age**, **Cabin**, and **Embarked** contain missing values.
- No duplicate records are present.
- Some features will require encoding before machine learning.
- Numerical features like **Fare** and **Age** should be checked for outliers.
- The dataset requires preprocessing before model training.

## 📝 Summary

In this notebook, we introduced the concept of Data Cleaning and inspected the Titanic dataset. We identified missing values, explored the dataset structure, reviewed data types, and understood why data preprocessing is an essential step before performing analysis or building machine learning models.